src/data/finset.lean

```diff
@@ -2132,6 +2176,79 @@ multiset.length_sort _
+
+lemma sorted_zero_eq_min' (s : finset α) (h : 0 < (s.sort (≤)).length) (H : s.nonempty) :
+  (s.sort (≤)).nth_le 0 h = s.min' H :=
+begin
+  let l := s.sort (≤),
+  apply le_antisymm,
+  { have : s.min' H ∈ l := (finset.mem_sort (≤)).mpr (s.min'_mem H),
+    obtain ⟨i, i_lt, hi⟩ : ∃ i (hi : i < l.length), l.nth_le i hi = s.min' H :=
+      list.mem_iff_nth_le.1 this,
+    rw ← hi,
+    exact list.nth_le_of_sorted_of_le (s.sort_sorted (≤)) (nat.zero_le i) },
+  { have : l.nth_le 0 h ∈ s := (finset.mem_sort (≤)).1 (list.nth_le_mem l 0 h),
+    exact s.min'_le H _ this }
+end
+
+lemma sorted_last_eq_max' (s : finset α) (h : (s.sort (≤)).length - 1 < (s.sort (≤)).length)
+  (H : s.nonempty) : (s.sort (≤)).nth_le ((s.sort (≤)).length - 1) h = s.max' H :=
+begin
+  let l := s.sort (≤),
+  apply le_antisymm,
+  { have : l.nth_le ((s.sort (≤)).length - 1) h ∈ s :=
+      (finset.mem_sort (≤)).1 (list.nth_le_mem l _ h),
+    exact s.le_max' H _ this },
+  { have : s.max' H ∈ l := (finset.mem_sort (≤)).mpr (s.max'_mem H),
+    obtain ⟨i, i_lt, hi⟩ : ∃ i (hi : i < l.length), l.nth_le i hi = s.max' H :=
+      list.mem_iff_nth_le.1 this,
+    rw ← hi,
+    have : i ≤ l.length - 1 := nat.le_pred_of_lt i_lt,
+    exact list.nth_le_of_sorted_of_le (s.sort_sorted (≤)) (nat.le_pred_of_lt i_lt) },
+end
+
+/-- Given a finset `s` of cardinal `k` in a linear order `α`, the map `mono_of_fin s h`
+is the increasing bijection between `fin k` and `s` as an `α`-valued map. Here, `h` is a proof that
+the cardinality of `s` is `k`. We use this instead of a map `fin s.card → α` to avoid
+casting issues in further uses of this function. -/
+def mono_of_fin (s : finset α) {k : ℕ} (h : s.card = k) (i : fin k) : α :=
+have A : (i : ℕ) < (s.sort (≤)).length, by simpa [h] using i.2,
+(s.sort (≤)).nth_le i A
+
+lemma bij_on_mono_of_fin (s : finset α) {k : ℕ} (h : s.card = k) :
+  set.bij_on (s.mono_of_fin h) set.univ ↑s :=
+begin
+  have A : ∀ j, j ∈ s ↔ j ∈ (s.sort (≤)) := λ j, by simp,
+  apply set.bij_on.mk,
+  { assume i hi,
+    simp only [mono_of_fin, set.mem_preimage, mem_coe, list.nth_le, A],
+    exact list.nth_le_mem _ _ _ },
+  { refine (strict_mono.injective (λ i j hij, _)).inj_on _,
+    exact list.pairwise_iff_nth_le.1 s.sort_sorted_lt _ _ _ hij },
+  { assume x hx,
+    simp only [mem_coe, A] at hx,
+    obtain ⟨i, il, hi⟩ : ∃ (i : ℕ) (h : i < (s.sort (≤)).length), (s.sort (≤)).nth_le i h = x :=
+      list.nth_le_of_mem hx,
+    simp [h] at il,
+    exact ⟨⟨i, il⟩, set.mem_univ _, hi⟩ }
+end
+
+/-- Given a finset `s` of cardinal `k` in a linear order `α`, the equiv `mono_equiv_of_fin s h`
+is the increasing bijection between `fin k` and `s` as an `s`-valued map. Here, `h` is a proof that
+the cardinality of `s` is `k`. We use this instead of a map `fin s.card → α` to avoid
+casting issues in further uses of this function. -/
+noncomputable def mono_equiv_of_fin (s : finset α) {k : ℕ} (h : s.card = k) : fin k ≃ s.to_set :=
```

Comment:
`s.to_set` is usually a coercion. For the subtype, I prefer `{x // x \in s}`. But I don't think we have a good convention here...

Reply:
I have switched to your notation. The better convention would be to have a coercion from `finset` to `Type*`, but this will have to wait for another PR.